### PR TITLE
Setup for benchmarks complete with travis

### LINKED ISSUE #1: Add benchmarks
Before optimizing the code any further we should add benchmarks in order to not do it blindly.

### PR DESCRIPTION
**Pull request by sn99**
fixes #1
~~I still need to figure out the best way for benchmarks and how they should be done but now it's just a matter of creating those functions~~, though running benches on travis could give wrong [results](https://github.com/bheisler/criterion.rs/issues/89) but anyhow I have added the script

### CODE REVIEW

**Comment by sn99** [comment]
@octavonce I have added benchmarks, I would like you to review them before I add more, Setting up was easy as the documentation examples seem to cover everything

Screenshot_2019-03-19 Job #3 1 - sn99 graphlib - Travis CI

End result in travis build for easy compare

**Comment by sn99** [comment]
@octavonce Benchmarks are complete

File: benches/benchmark.rs [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+// 10.remove(&mut self, id: &VertexId)
+// 11.remove_edge(&mut self, a: &VertexId, b: &VertexId)
+// 12.reserve(&mut self, additional: usize)
+// 13.retain(&mut self, fun: impl Fn(&T) -> bool)
+// 14.roots_count(&self) -> usize
+// 15.shrink_to_fit(&mut self)
+// 16.vertex_count(&self) -> usize
+fn bench_others(c: &mut Criterion) {
+    c.bench_function("add_edge", |b| {
+        b.iter(|| {
```
Comment (octavonce) [inline benches/benchmark.rs]: Can you separate the setup code from the benchmarked code like so:

```rust
c.bench_function("add_edge", |b| {
    let mut graph: Graph<usize> = Graph::new();

    let v1 = graph.add_vertex(1);
    let v2 = graph.add_vertex(2);
    
    b.iter(|| {
        graph.add_edge(&v1, &v2);
    })
});
```

File: benches/benchmark.rs [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+            let mut graph: Graph<usize> = Graph::new();
+
+            let v1 = graph.add_vertex(1);
+            let v2 = graph.add_vertex(2);
+            let v3 = graph.add_vertex(3);
+
+            graph.add_edge(&v1, &v2).unwrap();
+
+            assert!(graph.has_edge(&v1, &v2));
```
Comment (octavonce) [inline benches/benchmark.rs]: Please remove the asserts and separate the setup code.

**Review by octavonce — CHANGES_REQUESTED** [review]
Hi and thank you again for your contribution!

There are only two things to add:
* Can you please separate the benchmarked code from the setup code? I have added an example in the review.
* Can you add more benchmarks for the same method but with different vertex/edge counts? Such as for 10, 100, 1000, 5000, etc.? We should also benchmark really large graphs so we can optimize there as well.

File: benches/benchmark.rs (hunk 1 above)
Comment (sn99) [reply benches/benchmark.rs]: my bad

**Comment by sn99** [comment]
@octavonce What functions do you think should have more benchmarks ?

**Comment by octavonce** [comment]
@sn99 Well, it would be best if all of them did. Maybe except inserts and edge lookups because they are constant.

**Comment by sn99** [comment]
@octavonce I have completed benchmarks with variations of 10, 100, 500 and 1000 for all the benches (even for those who are constant for consistency sake)

**Comment by octavonce** [comment]
@sn99 Great! This looks good! I believe there are 2 more things and then we can merge:
* Can you change the name of the benchmarks to reflect the dimension of the graph? For example instead of "bfs 1", "bfs 2" it would be instead "bfs_10", "bfs_50", etc.
* Can you also add a benchmark to each method which has a stupendously large vertex/edge count? Such as greater than 5-10 million?

**Comment by sn99** [comment]
@octavonce renamed benchmarks, I also have added benchmarks with 10 million counts but they fail on travis and sometimes even locally so I have a conditional compilation flag `sbench` to use `cargo bench --features sbench` will run the benches with 10 milllion counts

**Review by octavonce — APPROVED** [review]
This is perfect! Merging now.

**Pull request merged**